### PR TITLE
Some Improvements

### PR DESCRIPTION
**Pull request by 2020karaulov2020**
Also it fix non active buttons like 'Clean' 'Optimize' 'Create Entity' 'Create BSP MODEL'. (it was uncomfortable)
Camera start origin is 0 0 0, but some maps not found at this coordinates. Move camera to first non worldspawn entity if presents.

### CODE REVIEW

**Comment by 2020karaulov2020** [comment]
Added working dir settings (for next updates, save / load files, etc)
Exit button (File->Exit, Very missing!)
Save first backup for edited map.

**Comment by 2020karaulov2020** [comment]
Added import/export lightmaps to LightMap Editor

**Comment by 2020karaulov2020** [comment]
Added export/import entity file and wad file.

File: src/editor/Gui.cpp [hunk 1]
```diff
@@ -2328,13 +2449,34 @@ void Gui::drawSettings() {
...
+						g_settings.workingdir[g_settings.workingdir.size() - 1] != '/')
+					{
+						g_settings.workingdir = g_settings.workingdir + "/";
+					}
+				}
+				else
+				{
+					g_settings.workingdir = "/";
+				}
+				sprintf(workingdir, "%s", g_settings.workingdir.c_str());
```
Comment (wootguy) [inline src/editor/Gui.cpp:2479]: Another reason to create a method that can joins paths, so that it doesn't matter if there are slashes or not in the setting.

File: src/editor/Gui.cpp [hunk 2]
```diff
@@ -3166,6 +3310,11 @@ void Gui::drawLightMapTool() {
 
 			if (showLightmapEditorUpdate)
 			{
+				logf("Clean previous lightmaps...\n");
+				for (int i = 0; i < MAXLIGHTMAPS; i++) {
+					sprintf(fileNam, "%s%s%s-%i.png", g_settings.gamedir.c_str(), g_settings.workingdir.c_str(), "lightmap", i);
+					removeFile(fileNam);
+				}
```
Comment (wootguy) [inline src/editor/Gui.cpp:3317]: This could erase someone's work, if they edited these lightmaps but clicked something else before importing them again. Safer to overwrite these if they exist, and only when the export button is pressed.

**Review by wootguy — CHANGES_REQUESTED** [review]

File: src/bsp/Bsp.cpp [hunk 3]
```diff
@@ -1826,9 +1826,26 @@ void Bsp::write(string path) {
...
+		if (!file.is_open()) {
+			logf("Failed to open backup file for writing:\n%s\n", path.c_str());
+			return;
+		}
+		logf("Writing backup to %s\n", (path + ".bak").c_str());
+
+		file.write(oldfile, len);
+		delete[] oldfile;
+	}
+
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:1844]: Ok

File: src/editor/Gui.cpp [hunk 4]
```diff
@@ -496,6 +496,116 @@ void Gui::drawMenuBar() {
...
+			if (ImGui::MenuItem("Entity file", NULL)) {
+				Bsp* map = app->getMapContainingCamera()->map;
+				if (map)
+				{
+					logf("Export entities: %s%s%s\n", g_settings.gamedir.c_str(), g_settings.workingdir.c_str(), "entities.ent");
+					ofstream entFile(g_settings.gamedir + g_settings.workingdir + "entities.ent", ios::out | ios::trunc);
+					entFile.write((const char*)map->lumps[LUMP_ENTITIES], map->header.lump[LUMP_ENTITIES].nLength);
+				}
+			}
+			if (ImGui::MenuItem("Wad file", NULL)) {
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:509]: Strange... I checked with some maps and it works, okay i try fix it.

File: src/editor/Gui.cpp [hunk 5]
```diff
@@ -496,6 +496,116 @@ void Gui::drawMenuBar() {
...
+					else
+					{
+						logf("No textures for export.\n");
+					}
+				}
+			}
+			ImGui::EndMenu();
+		}
+		if (ImGui::BeginMenu("Import")) {
+			if (ImGui::MenuItem("Entity file", NULL)) {
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:539]: Strange. And every export i got nore spaces after import)) okay I tried fix it

File: src/editor/Gui.cpp [hunk 6]
```diff
@@ -516,6 +626,7 @@ void Gui::drawMenuBar() {
 				logf("Check that the directories in the path exist, and that you have permission to write in them.\n");
 			}
 		}
+#endif
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:629]: :( okay....

File: src/editor/Gui.cpp (hunk 1 above)
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:2479]: Yes))

**Comment by wootguy** [comment]
> @wootguy how to remove LUMP_TEXTURES from map ? (I think need set to zero after wad exporting)

Resize the lump to 4 bytes and memset to 0.
http://hlbsp.sourceforge.net/index.php?content=bspdef#textures

File: src/editor/Renderer.cpp [hunk 7]
```diff
@@ -183,13 +186,19 @@ void AppSettings::load() {
...
 }
 
 void AppSettings::save() {
 	if (!dirExists(g_config_dir)) {
 		createDir(g_config_dir);
 	}
 
+	if (dirExists(g_settings.gamedir + g_settings.workingdir)) {
+		removeDir(g_settings.gamedir + g_settings.workingdir);
+	}
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:200]: May be,

File: src/util/util.cpp [hunk 8]
```diff
@@ -49,12 +67,18 @@ void debugf(const char* format, ...) {
...
+#elif defined(__cpp_lib_filesystem) 
+	return std::filesystem::exists(fileName) && !std::filesystem::is_directory(fileName);
+#else
 	if (FILE *file = fopen(fileName.c_str(), "r"))
 	{
 		fclose(file);
 		return true;
 	}
 	return false; 
+#endif
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp:81]: Create big code blocks is better?

[the remark this answers is not in the document]
Comment (wootguy) [reply src/util/util.cpp:81]: Having #ifdef blocks inside the function is fine, but I don't see the reason for adding the filesystem code here. Was something wrong with the FILE method? If so it'd better to completely replace this, ideally using something that doesn't require #ifdefs.

File: src/editor/Gui.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:629]: @wootguy it only for svencoop ? I can hide it for non sven game.

File: src/util/vectors.cpp [hunk 9]
```diff
@@ -217,6 +217,10 @@ vec3 vec3::flip() {
 	return vec3(x, z, -y);
 }
 
+vec3 vec3::abs() {
+	return vec3(fabs(x), fabs(y), fabs(z));
+}
+
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/vectors.cpp:223]: I think use it but not used) Ok))

File: src/editor/Gui.cpp (hunk 2 above)
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:3317]: Ok.

File: src/editor/Gui.cpp [hunk 10]
```diff
@@ -496,6 +496,116 @@ void Gui::drawMenuBar() {
...
+								map->add_texture(wadTex->szName, (byte *)imageData, wadTex->nWidth, wadTex->nHeight);
+
+								delete[] imageData;
+								delete wadTex;
+							}
+							for (int i = 0; i < app->mapRenderers.size(); i++) {
+								app->mapRenderers[i]->reloadTextures();
+							}
+						}
+						delete tmpWad;
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:598]: Ok moved

File: src/util/vectors.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/vectors.cpp:223]: Removed)

File: src/editor/Gui.cpp [hunk 11]
```diff
@@ -2156,6 +2275,8 @@ void Gui::drawSettings() {
 		ImGui::BeginChild("right pane content");
 		if (settingsTab == 0) {
 			ImGui::InputText("Game Directory", gamedir, 256);
+			ImGui::Text("Working Directory:");
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:2278]: @wootguy but in future it can be used for something do) Ok renamed.

File: src/editor/Renderer.cpp [hunk 12]
```diff
@@ -183,13 +186,19 @@ void AppSettings::load() {
 		resPaths.push_back("/svencoop_hd/");
 	}
 
+	if (!dirExists(g_settings.gamedir + g_settings.workingdir))
+		createDir(g_settings.gamedir + g_settings.workingdir);
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:190]: @wootguy  Now problem fixed. Can create folder with any deep.

File: src/editor/Renderer.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:200]: Okay

File: src/editor/Renderer.cpp [hunk 13]
```diff
@@ -1557,6 +1562,27 @@ void Renderer::addMap(Bsp* map) {
...
+		pickInfo.valid = true;
+		// Move camera to first entity with origin
+		for(auto const & ent : map->ents)
+		{
+			if (ent->getOrigin() != vec3())
+			{
+				cameraOrigin = ent->getOrigin();
+				break;
+			}
+		}
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:1584]: @wootguy some maps not located at 0 0 0, bspguy always start at 0 0 0 and it can too disorienting) Possible need detect center of map ?

File: src/util/util.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp:81]: @wootguy filesystem is cross platform and can work with '/' and '\\' slashes at same time, create recursive directories and etc.. But it require minimal c++14, i not remove old code for backward compatibility with c++11.

File: src/util/util.cpp [hunk 14]
```diff
@@ -7,6 +7,24 @@
...
+#include <Shlobj.h>
+#else 
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#endif
+
+
+#ifdef __cpp_lib_experimental_filesystem
+#include <experimental/filesystem>
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp:21]: Fixed now not needed define  _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING

File: src/util/util.cpp [hunk 15]
```diff
@@ -73,6 +97,43 @@ char * loadFile( const string& fileName, int& length)
...
+{
+#if defined(__cpp_lib_experimental_filesystem)
+	return std::experimental::filesystem::exists(fileName) && std::experimental::filesystem::remove(fileName);
+#elif defined(__cpp_lib_filesystem) 
+	return std::filesystem::exists(fileName) && std::filesystem::remove(fileName);
+#elif WIN32
+	return DeleteFile(fileName.c_str());
+#else 
+	return remove(fileName.c_str());
+#endif
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp:120]: https://github.com/wootguy/bspguy/pull/35#discussion_r581887438

**Comment by 2020karaulov2020** [comment]
In this part of fixes:
1. Now directory can be created recursive
2. Export/Import big code to functions
3. fixupPath function 
4. etc

**Comment by 2020karaulov2020** [comment]
> > @wootguy how to remove LUMP_TEXTURES from map ? (I think need set to zero after wad exporting)
> 
> Resize the lump to 4 bytes and memset to 0.
> http://hlbsp.sourceforge.net/index.php?content=bspdef#textures

This, LUMP_TEXTURES  used for save texture names, because of what there were problems with export) 

@wootguy you chan check all now ?

File: src/editor/Gui.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp]: yeah it's only for svencoop. I think it's not worth the extra code to conditionally hide this. Merging maps is the main purpose of this program. Maybe add in the tooltip that it's only for Sven Co-op.

File: src/editor/Gui.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp]: True, but even then the name "Working directory" could lead to confusion. It also means the directory from which you started the program.

File: src/editor/Renderer.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Renderer.cpp:1583]: Good point. Maybe a min/max check on the map vertices is good enough. As long as it's consistent.

File: src/util/util.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/util/util.cpp]: Ok those are really useful features. Creating recursive directories is a pain to do in a cross-platform way. Still, was this added to fix a bug, or can this be added later when it's needed? I don't see `/` and `\` being a problem since you can't share configs between linux and windows.

**Comment by wootguy** [comment]
> No function in bspguy to clear textures ? (remove data and set offsets to zero)
> It need to to use exported wad )

Oh right, there's a function for that. This function will delete the embedded textures, but not remove the texture names from the lump:
`Bsp::delete_embedded_textures`

File: src/editor/BspRenderer.cpp [hunk 16]
```diff
@@ -1503,6 +1503,11 @@ bool BspRenderer::pickPoly(vec3 start, vec3 dir, int hullIdx, PickInfo& pickInfo
 
 	start -= mapOffset;
 
+	if (!map || map->ents.size() == 0)
+	{
+		return false;
+	}
+
```
Comment (wootguy) [inline src/editor/BspRenderer.cpp:1510]: I think this would prevent you from being able to pick worldspawn polys? Just the check for `!map` should be good.

File: src/util/util.cpp [hunk 17]
```diff
@@ -7,6 +7,28 @@
...
+#define _SILENCE_EXPERIMENTAL_FILESYSTEM_DEPRECATION_WARNING
+#endif
+
+#ifdef __cpp_lib_filesystem
+#include <filesystem>
+namespace fs = std::filesystem;
+#define USE_FILESYSTEM
+#else 
+#include <experimental/filesystem>
+namespace fs = std::experimental::filesystem;
```
Comment (wootguy) [inline src/util/util.cpp:29]: nice to have fewer ifdefs with this change 👍

File: src/editor/Gui.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:580]: It's working now, and I see it's for embedded textures only. Can you rename the menu item to Export -> `Embedded Textures` ?

File: src/editor/Gui.cpp [hunk 18]
```diff
@@ -496,6 +496,116 @@ void Gui::drawMenuBar() {
...
+						entFile.read(data, datasize);
+						map->replace_lump(LUMP_ENTITIES, data, datasize);
+					}
+					else
+					{
+						logf("Error! No file!\n");
+					}
+				}
+			}
+			if (ImGui::MenuItem("Wad file", NULL)) {
```
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:612]: This still seems not to work with the latest code. I exported the WAD from this map, which has some embedded textures:
[duhhh.zip](https://github.com/wootguy/bspguy/files/6040454/duhhh.zip)
Then I exported the WAD, and edited the `duh_rock1` texture in wally, and saved the WAD.
image
After importing the WAD with bspguy, the rock textures doesn't change.

File: src/editor/Gui.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:591]: Seems to not do anything with the latest code either. How do you test it? Maybe I'm using it wrong.

File: src/util/util.cpp [hunk 19]
```diff
@@ -662,24 +708,201 @@ bool dirExists(const string& dirName_in)
...
+		return;
+	size_t start_pos = 0;
+	while ((start_pos = str.find(from, start_pos)) != std::string::npos) {
+		str.replace(start_pos, from.length(), to);
+		start_pos += to.length();
+	}
+}
+
+
+void fixupPath(std::string& path, FIXUPPATH_SLASH startslash, FIXUPPATH_SLASH endslash)
```
Comment (wootguy) [inline src/util/util.cpp:829]: This would eventually be removed in favor of something that can join paths that aren't "fixed". I want to get away from needing perfectly formatted paths for things to work. This code will likely be useful in creating a method that can join unfixed paths though.

**Review by wootguy — CHANGES_REQUESTED** [review]

File: src/editor/Gui.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:580]: Yes, if it's more understandable :)

File: src/editor/Gui.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:612]: @wootguy yes because internal textures with more priority or duplicated ? Possible need rename 'import' to 'merge with' and adding some fixes in code :)

File: src/editor/Gui.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: For example, I don't use sven coop and I don't even know what it is))) But what if this program become as full editor?)

File: src/editor/Gui.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: And what is the best name for the directory for do something operations?

File: src/util/util.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp]: WINAPI can work with '/' and '\' but linux only with '/'

File: src/editor/BspRenderer.cpp (hunk 16 above)
Comment (2020karaulov2020) [reply src/editor/BspRenderer.cpp:1510]: I got crash here when remove all entities from map )  See line https://github.com/wootguy/bspguy/blob/5c7e36d6bab6bd2a83c6f39e5fbc824c382c850e/src/editor/BspRenderer.cpp#L1515 )))

File: src/editor/Gui.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:612]: Oh ok, so importing doesn't replace textures, it appends to the existing set of textures.

I was thinking it would replace any textures that have the same name. It's ok to leave it the way it is now, but please add a tooltip explaining what it does.

File: src/editor/Gui.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp]: Makes sense. I wasn't thinking about other games when adding this. If you want, go ahead and add something to the tooltip to prevent confusion.

File: src/editor/Gui.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp]: "Working directory" 😄 

I'm having trouble coming up with a more generic name for this. "Import/Export directory" works for now. I'm questioning the need for having this at all. Usually things like this bring up a dialog that lets you choose where to save/load files.

File: src/editor/Gui.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: Okay ))

File: src/util/util.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/util/util.cpp]: I'm hesitant to use bleeding-edge features like this, especially if means older code needs to be maintained in parallel.

Linux users would have to type a \ by mistake to break something. I'm pretty sure most linux tools don't let you use \ anyway.

File: src/editor/BspRenderer.cpp (hunk 16 above)
Comment (wootguy) [reply src/editor/BspRenderer.cpp:1510]: Oh, that'll do it. I just tested and am still select faces when all entities are deleted. So, no issue here.

File: src/util/util.cpp (hunk 19 above)
Comment (2020karaulov2020) [reply src/util/util.cpp:829]: @wootguy imgui has file dialog for c++17 based at filesystem. Possible it can support c++14 also?
We can replace these function to file/directory open dialog, in future ?

File: src/editor/BspRenderer.cpp (hunk 16 above)
Comment (2020karaulov2020) [reply src/editor/BspRenderer.cpp:1510]: Strange...

**Comment by 2020karaulov2020** [comment]
@wootguy now all fixed ?)

**Comment by 2020karaulov2020** [comment]
@wootguy i found and fixed problem with entity import ) (render update)

File: src/editor/Gui.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:600]: Fixed in latest update)

File: src/editor/Renderer.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:1588]: I can't find it, just do 0,0,0 back :(

File: src/util/util.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp]: only c++17 has function to convert path directly to win/linux format make_preferred(), but i test and it not working in gcc, strange :)

[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/util/util.cpp]: For now, a "fixup" function is sufficient, i think)

File: src/editor/Gui.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: Fixed. Now it reeplace textures and append if no found.

File: src/editor/Gui.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: Okay now it Export/Import directory

File: src/bsp/Bsp.cpp [hunk 20]
```diff
@@ -2741,6 +2792,13 @@ int Bsp::add_texture(const char* name, byte* data, int width, int height) {
 		return -1;
 	}
 
+	BSPMIPTEX* oldtex = nullptr;
+
+	if (oldtex = find_embedded_texture(name))
```
Comment (wootguy) [inline src/bsp/Bsp.cpp:2797]: assignments inside conditions are hard to read (looks like a bug). I think this is more readable:

```
BSPMIPTEX* oldtex = find_embedded_texture(name);

if (oldtex)
```

File: src/bsp/Bsp.cpp [hunk 21]
```diff
@@ -2802,6 +2860,19 @@ int Bsp::add_texture(const char* name, byte* data, int width, int height) {
...
+		memcpy((byte*)oldtex + oldtex->nOffsets[0], mip[0], width * height);
+		memcpy((byte*)oldtex + oldtex->nOffsets[1], mip[1], (width >> 1) * (height >> 1));
+		memcpy((byte*)oldtex + oldtex->nOffsets[2], mip[2], (width >> 2) * (height >> 2));
+		memcpy((byte*)oldtex + oldtex->nOffsets[3], mip[3], (width >> 3) * (height >> 3));
+		memcpy((byte*)oldtex + (oldtex->nOffsets[3] + (width >> 3) * (height >> 3) + 2), palette, sizeof(COLOR3) * 256);
+		for (int i = 0; i < MIPLEVELS; i++) {
+			delete[] mip[i];
+		}
+		return 0;
+	}
```
Comment (wootguy) [inline src/bsp/Bsp.cpp:2874]: the new texture might be a different size, which could cause a crash or corruption.

File: src/editor/Gui.cpp [hunk 22]
```diff
@@ -2156,11 +2318,14 @@ void Gui::drawSettings() {
...
 		if (settingsTab == 0) {
 			ImGui::InputText("Game Directory", gamedir, 256);
+			ImGui::Text("Import/Export Directory:");
+			ImGui::InputText("(Relative to Game dir)", workingdir, 256);
 			if (ImGui::DragInt("Font Size", &fontSize, 0.1f, 8, 48, "%d pixels")) {
 				shouldReloadFonts = true;
 			}
 			ImGui::DragInt("Undo Levels", &app->undoLevels, 0.05f, 0, 64);
 			ImGui::Checkbox("Verbose Logging", &g_verbose);
+			ImGui::Checkbox("Make map backup", &g_settings.backUpMap);
```
Comment (wootguy) [inline src/editor/Gui.cpp:2328]: A tooltip here will be helpful:
```
		if (ImGui::IsItemHovered() && g.HoveredIdTimer > g_tooltip_delay) {
			ImGui::BeginTooltip();
			ImGui::TextUnformatted("Creates a backup of the BSP file when saving for the first time.");
		}
```

File: src/editor/Renderer.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Renderer.cpp:192]: can this be created when it's needed instead? If the folder is deleted while the program is open then files can't be exported.

**Comment by 2020karaulov2020** [comment]
@wootguy now no problem with import/export ?

File: src/editor/Renderer.cpp [hunk 23]
```diff
@@ -504,22 +514,8 @@ void Renderer::renderLoop() {
...
-
-			for (int i = 0; i < mapRenderers.size(); i++) {
-				mapRenderers[i]->pointEntRenderer = pointEntRenderer;
-				mapRenderers[i]->preRenderEnts();
-				if (reloadingGameDir) {
-					mapRenderers[i]->reloadTextures();
-				}
-			}
-
+			reloadFgds();
```
Comment (wootguy) [inline src/editor/Renderer.cpp:517]: Rename to `postLoadFgds()`. This block is entered after FGDs have loaded.

**Review by wootguy — CHANGES_REQUESTED** [review]
Import/Export seems to work now with WADs. I think the old `Import -> WAD` naming makes sense now, since it's replacing textures instead of appending new ones. A tooltip there would help to explain what exactly is happening. Something like:
"Embeds textures from `workingdir`/`mapname`.wad. Existing textures with the same names will be replaced."

Importing entities still isn't working for me, following the instructions I commented earlier. I can take a look at the code later to see what's wrong.

File: src/bsp/Bsp.cpp (hunk 21 above)
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:2874]: Yes I can rename new texture to another  if w/h different

File: src/editor/BspRenderer.cpp (hunk 16 above)
Comment (wootguy) [reply src/editor/BspRenderer.cpp:1510]: Right, I meant I tested it with your fix and I was wrong that there would be a problem with selecting faces when there's 0 ents.

File: src/util/util.cpp (hunk 19 above)
Comment (wootguy) [reply src/util/util.cpp:829]: If it compiles in VS 2019 then I'm fine with adding that in later. VS seems to be slowest to adopt new features.

File: src/bsp/Bsp.cpp (hunk 21 above)
Comment (wootguy) [reply src/bsp/Bsp.cpp:2874]: Would make more sense to replace the existing one, but this code is kind of a pain to work with so renaming is fine too.

File: src/editor/Renderer.cpp (hunk 23 above)
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp:517]: ok

File: src/editor/Gui.cpp (hunk 22 above)
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:2328]: ok

File: src/bsp/Bsp.cpp (hunk 21 above)
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:2874]: Ok

File: src/bsp/Bsp.cpp (hunk 20 above)
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:2797]: ok!

**Comment by 2020karaulov2020** [comment]
@wootguy you don't know how to create empty map with possible to load in game ? (I tried create maps without compilers, directly from bspguy 🤣 🤣 🤣  )

check main.cpp:132

File: src/util/util.cpp (hunk 19 above)
Comment (2020karaulov2020) [reply src/util/util.cpp:829]: And https://github.com/AirGuanZ/imgui-filebrowser for c++17

Comment (2020karaulov2020) [reply src/util/util.cpp:829]: @wootguy i can't watch notices in github, and can skip some messages from you. You has discord or something like ?

**Comment by 2020karaulov2020** [comment]
I success added -testnew command to create working map with one cube )))
But i see one bad triangle: https://www.youtube.com/watch?v=8XcUFVEsHuY  (inserted bsp model via cycler_sprite)

@wootguy you can check now this pull request ?

**Comment by wootguy** [comment]
There's too much going on in this pull request. Can you split it up into to smaller pieces next time? Like one PR for each feature:
* Import/export WAD
* Import/export entities
* Import/export lightmaps
* create empty BSP
* other improvements

It's getting hard to follow and see all the comments here.

**Comment by wootguy** [comment]
> @wootguy you don't know how to create empty map with possible to load in game ? (I tried create map without compilers, directly from bspguy 🤣 🤣 🤣 )

Haven't tried it before, maybe there's something special needed for creating the worldspawn BSP model.

File: src/bsp/Bsp.cpp [hunk 24]
```diff
@@ -25,9 +26,25 @@ Bsp::Bsp() {
...
+			break;
+		case LUMP_LIGHTING:
+			header.lump[i].nLength = 204800;
+			break;
+		default:
+			header.lump[i].nLength = 0;
+			break;
+		}
+	
+		lumps[i] = new byte[header.lump[i].nLength + 4];
```
Comment (wootguy) [inline src/bsp/Bsp.cpp:43]: why +4 here?

File: src/bsp/Bsp.cpp (hunk 21 above)
Comment (wootguy) [reply src/bsp/Bsp.cpp:2903]: still a problem with the latest code

File: src/main.cpp [hunk 25]
```diff
@@ -123,12 +125,51 @@ void hideConsoleWindow() {
...
+			thisFaces[i].nLightmapOffset = 768 * (i + 1);
+		}
+		// Update all lumps
+		newBsp->update_ent_lump();
+		newBsp->update_lump_pointers();
+		// Render map
+		renderer.addMap(newBsp);
+	}
+	else 
+		renderer.addMap(new Bsp(map));
```
Comment (wootguy) [inline src/main.cpp:172]: can you leave this out of the PR, or make a separate PR for this? I want to focus on fixing what's here first.

File: src/editor/Gui.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:604]: still not doing anything for me. Maybe this should go in its own PR. I said I would debug this but don't have energy for it at the moment.

**Review by wootguy — CHANGES_REQUESTED** [review]

File: src/util/util.cpp (hunk 19 above)
Comment (wootguy) [reply src/util/util.cpp:829]: Yeah, in discord I'm w00tguy#9718

I sometimes am missing messages here, too.

File: src/editor/Gui.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:604]: Strange i export entities then remove one, one replace, press import. And entity removed and of other entity origin updated.

File: src/bsp/Bsp.cpp (hunk 21 above)
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:2903]: Sorry i skip your message)

File: src/bsp/Bsp.cpp (hunk 24 above)
Comment (2020karaulov2020) [reply src/bsp/Bsp.cpp:43]: test, need remove

File: src/main.cpp (hunk 25 above)
Comment (2020karaulov2020) [reply src/main.cpp:172]: Yes, partially.

File: src/editor/Gui.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:604]: alright, it works if I follow those steps. This is what I've been doing, which doesn't work:
* Delete entity
* Export -> ents
* File -> Relaod
* Import -> ents

and the entity that should've been deleted is still there.

**Comment by 2020karaulov2020** [comment]
> There's too much going on in this pull request. Can you split it up into to smaller pieces next time? Like one PR for each feature:
> 
> * Import/export WAD
> * Import/export entities
> * Import/export lightmaps
> * create empty BSP
> * other improvements
> 
> It's getting hard to follow and see all the comments here.

Now it can be problematic, i can create import/export pull request and create test cube bsp pull request.

File: src/editor/Renderer.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Renderer.cpp]: in new pull it fixed
